compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -1115,51 +1118,31 @@ pub fn typeid_for_instance<'tcx>(
         instance.args = strip_receiver_auto(tcx, instance.args)
     }
 
+    if let Some(impl_id) = tcx.impl_of_method(instance.def_id())
+        && let Some(trait_ref) = tcx.impl_trait_ref(impl_id)
+    {
+        // Trait methods will have a Self polymorphic parameter, where the concreteized
+        // implementatation will not. We need to walk back to the more general trait method
+        let trait_ref = trait_ref.instantiate(tcx, instance.args);
+        let invoke_ty = trait_object_ty(tcx, ty::Binder::dummy(trait_ref));
+        let method_id = tcx
+            .impl_item_implementor_ids(impl_id)
+            .items()
+            .filter_map(|(trait_method, impl_method)| {
+                (*impl_method == instance.def_id()).then_some(*trait_method)
+            })
+            .min()
+            .unwrap();
+        instance.def = ty::InstanceDef::Virtual(method_id, 0);
```

Comment:
Doesnt this mean we won't catch transmuting rcvrs if they end up with the same trait object representation

Reply:
To get the right type on the rewritten instance, I need to replace the `def_id` with the `method_id` I've looked up. I thought the easiest way to reason about it was to make the call-site instance (which would be a `Virtual`) match the callee instance according to the alias set computation, as that would mean they always generate the same value.

The `0` is because we don't currently consume that. I could theoretically call `.vtable_entries()` and look up the ID if we want it to be completely matching.

Would you prefer I use a different `InstanceDef`, or try to preserve the existing one, or just document what's going on here in a comment?

Reply:
A comment about what this is doing inline + an response to the question above on this thread would help

Reply:
Sorry, started typing that response before your second comment appeared on my end - I'm not sure what you mean by not catching transmuting receivers, my existing comment doesn't explain or respond to that.

Reply:
Yes I know. I'm asking for your explanation for the first part to make it into the code for posterity, and I was asking the thing about self rcvrs on the side.

Reply:
I've added the comment you asked for, but I'm not sure I understand what you're asking about with regards to catching transmuting receivers.

Reply:
I'll try to write up an example of an expected SIGILL that doesn't end up triggering after this PR. 

Essentially this is projecting all self types to object versions of their self type. This means that if two &self types from different impl end up projecting to the same &dyn Trait, we won't end up detecting a mismatch in them.

Reply:
That was already the case - the code this deletes already rewrites `&self` to `&dyn Self`. This just makes it much more general, and handle associated types and non `&self`/`&mut self` receivers.

I don't think we could detect that in the general case anyways - the callsite only knows that the pointer it has is a pointer to something that implements that trait. It doesn't know which concrete type, so it can't enforce which impl it's going to.

Reply:
The thing this does leave incorrect is function pointers to trait methods - but those are already completely broken. Both before and after this patch, if I have
```
trait Foo {
  fn f(&self);
}
struct FooImpl;
impl Foo for FooImpl {
  fn f(&self) {}
}
```

`FooImpl::f` has a label of `fn(&dyn Foo)`. This means that:
1. If you call it at a callsite expecting a `fn(&FooImpl)`, it will SIGILL
2. If you call it at a callsite expecting a `fn(&dyn Foo)`, it will incorrectly permit it.

Fixing this while keeping vtables working requires a shim in KCFI (though I might be able to do it with `Reify` rather than the custom one I had before), and either a shim or multiple type attachment in CFI.

Reply:
Multiple types would be cool

Reply:
Do you want me to stack that into the same change? I was attempting to make my changes smaller based on previous feedback. I don't think this change as it stands breaks anything (other than potentially the rebasing below which I'll address), so I was trying to avoid stacking too many fixes together.

Reply:
This is definitely a follow-up